### DISCUSSION TITLE
Modifica stato pagamento "Issued Document" e "Received Documents"

### DISCUSSION THREAD

**Discussion by overiver1** [fattureincloud/api#239] [Q&A] [answered]
Salve sto effettuando la migrazione da V1 a V2 utilizzando il token permanente tutto ok finora ma non riesco a capire come poter modificare lo stato di pagamento di una fattura in entrata o in uscita.
Nell'app ho settato il controllo completo per  fatture in entrata e in uscita
Faccio riferimento alle **API reference** e qui trovo il blocco "**Modify Received Document**" e "**Modify Issued Document**" ma anche nelle Reference inserendo id della company e quello di un documento preso dalla lista ricevo il messaggio "risorsa non trovata" e non trovo campi dove inserire le voci da modificare 
Per i test sto utilizzando Curl con metodo PUT, ho provato ad inserire: 
$data = '{"data":{"payments_list":[{"status":"not_paid"}]}}';
ma ricevo: 400 Bad Request

**Comment by emajo** [comment, marked as answer]
Ciao @overiver1, gli errori che stai riportando sono molto strani in quanto risorsa non trovata dovrebbe restituire un 404 mentre poi ti viene restituito un 400, quindi la risorsa viene trovata, da api reference mi sembra molto strano che ti restituista un errore del genere a meno che l'id non sia effettivamente errato.

Per quanto riguarda il body da passare nella chiamata di modifica documento non basta passare solamente il campo status ma va ripassato completamente tutto il payment, per esempio se allo stato attuale il campo payments items è popolato così:
```
...
"payments_list": [
            {
                "amount": 18.75,
                "due_date": "2023-02-22",
                "status": "not_paid",
            }
        ]
...
```
per essere segnato come pagato dovrà essere passato un body del genere:
```
"data": {
     "payments_list": [
            {
                "amount": 18.75,
                "due_date": "2022-02-22",
                "paid_date": "2022-12-24",
                "status": "paid",
                "payment_account": {
                    "id": 999,
                }
            }
        ]
}
```

Ci faccia sapere se riesce a risolvere,
grazie,
Emanuele.

**Comment by overiver1** [comment]
Buongiorno, fatto un test ma non va:
questo  è quello che inserisco nel body

`$data  = array(
  "data"=>array([
    "payments_list"=>array(
	   "amount"=>1020,
	   "due_date"=>"2023-01-24",
	   "paid_date"=>"2023-01-24",
       "status"=>"paid",
	   "payment_account"=>array(
            "id"=>569803,	
		)
    )
  ])
);`

**Reply by overiver1** [reply]
nella risposta leggo  l'aggiornamento: "updated_at":"2023-01-24 11:13:26" ma lo stato rimane inalterato "status":"not_paid

**Reply by valmoz** [reply]
Buongiorno @overiver1 
per favore, ci invii il JSON della richiesta. 

Grazie
Mauro

**Comment by overiver1** [comment]
Grazie per il supporto 
e questo è il code con cui effettuo il PUT:

<?php
$url = "https://api-v2.fattureincloud.it/c/6076/issued_documents/2564...............";

$curl = curl_init();
curl_setopt($curl, CURLOPT_URL, $url);
curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "PUT");
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);

$headers = array(
   "Accept: application/json",
   "Authorization: Bearer a/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9....................................."
);
curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);

$data  = array(
  "data"=>array([
    "payments_list"=>array(
       "amount"=>1020,
       "due_date"=>"2023-01-24",
       "paid_date"=>"2023-01-24",
       "status"=>"paid",
       "payment_account"=>array(
       "id"=>569803    
        )
    )
  ])
);    

curl_setopt($curl, CURLOPT_POSTFIELDS, http_build_query($data));

curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

$resp = curl_exec($curl);
curl_close($curl);
var_dump($resp);
?>

**Reply by emajo** [reply]
Ciao @overiver1 sembrerebbe esserci un errore nella costruzione di data, così dovrebbe funzionare:
```
$data = [
    "data" => [
        "payments_list" => [
            "amount" => 1020,
            "due_date" => "2023-01-24",
            "paid_date" => "2023-01-24",
            "status" => "paid",
            "payment_account" => ["id" => 569803],
        ],
    ],
];
```

**Reply by valmoz** [reply]
Buongiorno @overiver1 
non effettueremo il debug del suo codice custom.

Le chiedo gentilmente di stampare il campo $data in formato JSON come descritto qui: [JSON Encode](https://www.php.net/manual/en/function.json-encode.php)

Comunque, dai log mi risulta un body totalmente diverso da quello che sembrerebbe essere generato dal suo codice (ho censurato i valori):
```
{
    "data": {
        "entity": {
            "id": 205000,
            "name": "Ms PacXXX",
            "vat_number": "081XXX",
            "tax_code": "081XXX",
            "address_street": "Via dei CarrXXX",
            "address_postal_code": "84121",
            "address_city": "Salerno",
            "address_province": "SA",
            "address_extra": "",
            "country": "Italia",
            "certified_email": "",
            "ei_code": "SUBXXX",
            "type": null
        }
    }
}
```
(ovviamente, non avendo la certezza del company id spero di aver preso le chiamate giuste)

**Reply by valmoz** [reply]
Ne approfitto per suggerirle di usare la nostra [PHP SDK](https://developers.fattureincloud.it/docs/sdks/php-sdk)

**Reply by overiver1** [reply]
Salve, 
Grazie, così sembra funzionare:

`$data = array(
    "data" =>  array(
        "payments_list" => array([
            "amount" => 243.02,
            "due_date" => "2023-01-11",
            "paid_date" => "2023-01-24",
            "status" => "paid",
            "payment_account" => array("id" => 572xxx),
      ]),
    ),
);`

Approfitto per chiedere non si può omettere in qualche modo il valore "payment_account" => array("id" => 572xxx),

**Reply by emajo** [reply]
Il payment_account risulta obbligatorio in caso di status = paid.

**Reply by overiver1** [reply]
ok, Grazie per il prezioso supporto

**Reply by dcesca** [reply]
@valmoz buongiorno, io sto usando l'SDK e mi servirebbe proprio fare l'operazione di settaggio dello stato pagato.
Mi potresti indirizzare alla documentazione su dove si può fare questa operazione?

**Reply by dcesca** [reply]
@valmoz trovato https://developers.fattureincloud.it/docs/guides/invoice-creation/#six-step-six-payments-list

**Comment by dcesca** [comment]
@valmoz ciao, un chiarimento. Se io ho un received document con più di un pagamento (magari la classica fattura a 30/60/90) e io devo mettere a pagato uno solo di questi, devo passare tutto l’elenco dei pagamenti con solo quello da modificare cambiato? O posso mandare solo quello cambiato e FiC fa il match con la due_date e importo?

**Reply by emajo** [reply]
Ciao @dcesca, le chiedo la cortesia di aprire una nuova discussion specifica a questa tematica in quanto potrebbe risultare utile anche ad altri utenti.

Grazie,
Emanuele.
